File: AssistantPlus/Application/Utility/Finalizer.cs
#pragma warning disable 0,
// ReSharper disable

using Twinning.AssistantPlus;

namespace Twinning.AssistantPlus.Utility {

	public partial class Finalizer : IDisposable, IAsyncDisposable {

		#region constructor

		private Func<Task> mAction;

		// ----------------

		public Finalizer (
			Func<Task> action
		) {
			this.mAction = action;
			return;
		}

		#endregion

		#region implement IDisposable

		public void Dispose (
		) {
			this.mAction();
			GC.SuppressFinalize(this);
			return;
		}

		#endregion

		#region implement IAsyncDisposable

		public async ValueTask DisposeAsync (
		) {
			await this.mAction();
			GC.SuppressFinalize(this);
			return;
		}

		#endregion

	}

}

File: module/assistant/lib/utility/finalizer.dart
import '/common.dart';

// ----------------

class Finalizer {

  // #region constructor

  Future<Void> Function() _action;

  Boolean _disposed;

  // ----------------

  Finalizer(
    Future<Void> Function() action,
  ) :
    this._action = action,
    this._disposed = false;

  // #endregion

  // #region dispose

  Future<Void> dispose(
  ) async {
    assertTest(!this._disposed);
    this._disposed = true;
    await this._action();
    return;
  }

  // #endregion

}
